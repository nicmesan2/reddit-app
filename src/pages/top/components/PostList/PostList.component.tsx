import React from 'react'
import styled from 'styled-components'
import { CSSTransition, TransitionGroup } from 'react-transition-group'
import { ImageInterface, PostInterface } from '../../topList.types'
import { Post } from '../../../../components'

const Message = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100%;
`

interface TopListInterface {
  isLoading?: boolean
  posts: PostInterface[]
  onClick: (postId?: string) => {}
  onDelete: (postId?: string) => {}
  onSave: (postId?: string, image?: ImageInterface, imageBookmarked?: boolean) => {}
}

const PostList: React.FC<TopListInterface> = ({ isLoading = false, posts, onClick, onDelete, onSave }) => {
  if (isLoading) {
    const mockedData = Array(12).fill({ isLoading: true })

    return (
      <div>
        {mockedData.map((placeholder, i) => (
          <Post key={i} isLoading />
        ))}
      </div>
    )
  }

  if (posts.length === 0) {
    return <Message>No posts found :(</Message>
  }

  return (
    <TransitionGroup>
      {posts.map((post) => (
        <CSSTransition timeout={200} classNames="post" key={post.id}>
          <Post onSave={onSave} onPostClick={onClick} onDeleteClick={onDelete} {...post} />
        </CSSTransition>
      ))}
    </TransitionGroup>
  )
}

export default PostList
