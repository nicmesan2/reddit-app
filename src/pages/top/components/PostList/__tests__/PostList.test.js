import React from 'react'
import 'jest-styled-components'
import { render, cleanup } from '@testing-library/react'
import PostList from '../PostList.component'

afterEach(cleanup)

jest.mock('react-transition-group', () => ({
  TransitionGroup: props => <div>{props.children}</div>,
  CSSTransition: (props) => <div>{props.children}</div>
}))

const MOCK_POST = {
  onSave: () => {},
  id: 'hqb54s',
  createdTime: 1594626450,
  title: 'The music just sounds good',
  author: 'gamerking586',
  thumbnail: 'https://b.thumbs.redditmedia.com/SfeqzrTADvsk0JxxSg9RPQPMMr-0eYNuF6hoMurYerQ.jpg',
  commentsNumber: 1156,
  clicked: false,
  image: {
    src: 'https://preview.redd.it/rc6bq0lmxka51.jpg?auto=webp&s=25e97d99885da17754fdd65b7fcd984f9537eeb4',
    thumbnail: 'https://b.thumbs.redditmedia.com/SfeqzrTADvsk0JxxSg9RPQPMMr-0eYNuF6hoMurYerQ.jpg',
    thumbnailHeight: 140,
    thumbnailWidth: 140
  }
}

test('Renders', () => {
  const { asFragment } = render(<PostList isLoading />)
  expect(asFragment()).toMatchSnapshot()
})

test('Renders placeholders when in loading state', () => {
  const { getAllByTestId } = render(<PostList isLoading />)
  expect(getAllByTestId('Placeholder')).toBeTruthy()
})

test('Renders when posts length is cero', () => {
  const { getByText } = render(<PostList posts={[]} />)

  expect(getByText('No posts found', { exact: false })).toBeTruthy()
})

test('Renders posts', async () => {
  const mockedPosts = Array(3).fill().map((_, i) => ({
    ...MOCK_POST,
    title: `Post ${i}`,
    id: i
  }))
  
  const { getAllByTestId } = render(
    <PostList posts={mockedPosts} />
  )
  
  const postTitles = await getAllByTestId('Post_title').map(titleElement => titleElement.textContent)
  const mockedPostTitles = mockedPosts.map((post) => post.title)
  expect(postTitles).toEqual(mockedPostTitles)
})
