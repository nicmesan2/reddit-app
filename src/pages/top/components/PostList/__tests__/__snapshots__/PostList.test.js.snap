// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders 1`] = `
<DocumentFragment>
  .c1 {
  display: block;
  width: 45px;
  height: 16px;
  border-radius: 30px;
  -webkit-animation-duration: 2s;
  animation-duration: 2s;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
  -webkit-animation-name: placeholderAnimatedBackground;
  animation-name: placeholderAnimatedBackground;
  -webkit-animation-timing-function: linear;
  animation-timing-function: linear;
  background: #eee;
  background: linear-gradient(to right,#eee 10%,#ddd 20%,#eee 35%);
  background-size: 800px 100px;
}

.c0 {
  cursor: pointer;
  border-top: thin solid #ccc;
  border-bottom: thin solid #ccc;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  padding: 8px;
  margin-bottom: -1px;
  background-color: #ffffff;
}

.c0:hover {
  border: thin solid black;
  z-index: 1;
}

.c2 {
  overflow: hidden;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-left: 8px;
}

.c3 {
  margin: 5px 0;
  color: #787c7e;
  font-size: 12px;
  font-weight: 400;
  line-height: 16px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div>
    <div
      class="c0"
    >
      <div
        class="c1"
        data-testid="Placeholder"
        style="width: 80px; height: 60px; animation-delay: 0.22s;"
      />
      <div
        class="c2"
      >
        <div
          class="c1"
          data-testid="Placeholder"
          style="width: 300px; animation-delay: 0.22s;"
        />
        <div
          class="c3"
        >
          <div
            class="c1"
            data-testid="Placeholder"
            style="width: 200px; animation-delay: 0.22s;"
          />
        </div>
        <div
          class="c4"
        />
      </div>
      <div
        class="c5"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-angle-right fa-w-8 "
          data-icon="angle-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style="color: rgb(255, 69, 0);"
          viewBox="0 0 256 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    <div
      class="c0"
    >
      <div
        class="c1"
        data-testid="Placeholder"
        style="width: 80px; height: 60px; animation-delay: 0.22s;"
      />
      <div
        class="c2"
      >
        <div
          class="c1"
          data-testid="Placeholder"
          style="width: 300px; animation-delay: 0.22s;"
        />
        <div
          class="c3"
        >
          <div
            class="c1"
            data-testid="Placeholder"
            style="width: 200px; animation-delay: 0.22s;"
          />
        </div>
        <div
          class="c4"
        />
      </div>
      <div
        class="c5"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-angle-right fa-w-8 "
          data-icon="angle-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style="color: rgb(255, 69, 0);"
          viewBox="0 0 256 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    <div
      class="c0"
    >
      <div
        class="c1"
        data-testid="Placeholder"
        style="width: 80px; height: 60px; animation-delay: 0.22s;"
      />
      <div
        class="c2"
      >
        <div
          class="c1"
          data-testid="Placeholder"
          style="width: 300px; animation-delay: 0.22s;"
        />
        <div
          class="c3"
        >
          <div
            class="c1"
            data-testid="Placeholder"
            style="width: 200px; animation-delay: 0.22s;"
          />
        </div>
        <div
          class="c4"
        />
      </div>
      <div
        class="c5"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-angle-right fa-w-8 "
          data-icon="angle-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style="color: rgb(255, 69, 0);"
          viewBox="0 0 256 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    <div
      class="c0"
    >
      <div
        class="c1"
        data-testid="Placeholder"
        style="width: 80px; height: 60px; animation-delay: 0.22s;"
      />
      <div
        class="c2"
      >
        <div
          class="c1"
          data-testid="Placeholder"
          style="width: 300px; animation-delay: 0.22s;"
        />
        <div
          class="c3"
        >
          <div
            class="c1"
            data-testid="Placeholder"
            style="width: 200px; animation-delay: 0.22s;"
          />
        </div>
        <div
          class="c4"
        />
      </div>
      <div
        class="c5"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-angle-right fa-w-8 "
          data-icon="angle-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style="color: rgb(255, 69, 0);"
          viewBox="0 0 256 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    <div
      class="c0"
    >
      <div
        class="c1"
        data-testid="Placeholder"
        style="width: 80px; height: 60px; animation-delay: 0.22s;"
      />
      <div
        class="c2"
      >
        <div
          class="c1"
          data-testid="Placeholder"
          style="width: 300px; animation-delay: 0.22s;"
        />
        <div
          class="c3"
        >
          <div
            class="c1"
            data-testid="Placeholder"
            style="width: 200px; animation-delay: 0.22s;"
          />
        </div>
        <div
          class="c4"
        />
      </div>
      <div
        class="c5"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-angle-right fa-w-8 "
          data-icon="angle-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style="color: rgb(255, 69, 0);"
          viewBox="0 0 256 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    <div
      class="c0"
    >
      <div
        class="c1"
        data-testid="Placeholder"
        style="width: 80px; height: 60px; animation-delay: 0.22s;"
      />
      <div
        class="c2"
      >
        <div
          class="c1"
          data-testid="Placeholder"
          style="width: 300px; animation-delay: 0.22s;"
        />
        <div
          class="c3"
        >
          <div
            class="c1"
            data-testid="Placeholder"
            style="width: 200px; animation-delay: 0.22s;"
          />
        </div>
        <div
          class="c4"
        />
      </div>
      <div
        class="c5"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-angle-right fa-w-8 "
          data-icon="angle-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style="color: rgb(255, 69, 0);"
          viewBox="0 0 256 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    <div
      class="c0"
    >
      <div
        class="c1"
        data-testid="Placeholder"
        style="width: 80px; height: 60px; animation-delay: 0.22s;"
      />
      <div
        class="c2"
      >
        <div
          class="c1"
          data-testid="Placeholder"
          style="width: 300px; animation-delay: 0.22s;"
        />
        <div
          class="c3"
        >
          <div
            class="c1"
            data-testid="Placeholder"
            style="width: 200px; animation-delay: 0.22s;"
          />
        </div>
        <div
          class="c4"
        />
      </div>
      <div
        class="c5"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-angle-right fa-w-8 "
          data-icon="angle-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style="color: rgb(255, 69, 0);"
          viewBox="0 0 256 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    <div
      class="c0"
    >
      <div
        class="c1"
        data-testid="Placeholder"
        style="width: 80px; height: 60px; animation-delay: 0.22s;"
      />
      <div
        class="c2"
      >
        <div
          class="c1"
          data-testid="Placeholder"
          style="width: 300px; animation-delay: 0.22s;"
        />
        <div
          class="c3"
        >
          <div
            class="c1"
            data-testid="Placeholder"
            style="width: 200px; animation-delay: 0.22s;"
          />
        </div>
        <div
          class="c4"
        />
      </div>
      <div
        class="c5"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-angle-right fa-w-8 "
          data-icon="angle-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style="color: rgb(255, 69, 0);"
          viewBox="0 0 256 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    <div
      class="c0"
    >
      <div
        class="c1"
        data-testid="Placeholder"
        style="width: 80px; height: 60px; animation-delay: 0.22s;"
      />
      <div
        class="c2"
      >
        <div
          class="c1"
          data-testid="Placeholder"
          style="width: 300px; animation-delay: 0.22s;"
        />
        <div
          class="c3"
        >
          <div
            class="c1"
            data-testid="Placeholder"
            style="width: 200px; animation-delay: 0.22s;"
          />
        </div>
        <div
          class="c4"
        />
      </div>
      <div
        class="c5"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-angle-right fa-w-8 "
          data-icon="angle-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style="color: rgb(255, 69, 0);"
          viewBox="0 0 256 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    <div
      class="c0"
    >
      <div
        class="c1"
        data-testid="Placeholder"
        style="width: 80px; height: 60px; animation-delay: 0.22s;"
      />
      <div
        class="c2"
      >
        <div
          class="c1"
          data-testid="Placeholder"
          style="width: 300px; animation-delay: 0.22s;"
        />
        <div
          class="c3"
        >
          <div
            class="c1"
            data-testid="Placeholder"
            style="width: 200px; animation-delay: 0.22s;"
          />
        </div>
        <div
          class="c4"
        />
      </div>
      <div
        class="c5"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-angle-right fa-w-8 "
          data-icon="angle-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style="color: rgb(255, 69, 0);"
          viewBox="0 0 256 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    <div
      class="c0"
    >
      <div
        class="c1"
        data-testid="Placeholder"
        style="width: 80px; height: 60px; animation-delay: 0.22s;"
      />
      <div
        class="c2"
      >
        <div
          class="c1"
          data-testid="Placeholder"
          style="width: 300px; animation-delay: 0.22s;"
        />
        <div
          class="c3"
        >
          <div
            class="c1"
            data-testid="Placeholder"
            style="width: 200px; animation-delay: 0.22s;"
          />
        </div>
        <div
          class="c4"
        />
      </div>
      <div
        class="c5"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-angle-right fa-w-8 "
          data-icon="angle-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style="color: rgb(255, 69, 0);"
          viewBox="0 0 256 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    <div
      class="c0"
    >
      <div
        class="c1"
        data-testid="Placeholder"
        style="width: 80px; height: 60px; animation-delay: 0.22s;"
      />
      <div
        class="c2"
      >
        <div
          class="c1"
          data-testid="Placeholder"
          style="width: 300px; animation-delay: 0.22s;"
        />
        <div
          class="c3"
        >
          <div
            class="c1"
            data-testid="Placeholder"
            style="width: 200px; animation-delay: 0.22s;"
          />
        </div>
        <div
          class="c4"
        />
      </div>
      <div
        class="c5"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-angle-right fa-w-8 "
          data-icon="angle-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          style="color: rgb(255, 69, 0);"
          viewBox="0 0 256 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
